File: daily/Sliding Window/992-Subarrays-with-k-different-integers.go
package daily

func subarraysWithKDistinct(nums []int, k int) int {
	return atMost(nums, k) - atMost(nums, k-1)
}

func atMost(nums []int, k int) int {
	count := make(map[int]int)
	left := 0
	res := 0
	n := len(nums)
	for right := 0; right < n; right++ {
		count[nums[right]]++
		for len(count) > k {
			count[nums[left]]--
			if count[nums[left]] == 0 {
				delete(count, nums[left])
			}
			left++
		}
		res += right - left + 1
	}
	return res
}

File: daily/Sliding Window/992-Subarrays-with-k-different-integers.py
from collections import defaultdict

class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        return self.atMost(nums, k) - self.atMost(nums, k - 1)
    def atMost(self, nums: List[int], k: int) -> int:
        count = defaultdict(int)
        left = res = 0
        n = len(nums)
        for right in range(n):
            count[nums[right]] += 1
            while len(count) > k:
                count[nums[left]] -= 1
                if count[nums[left]] == 0:
                    del count[nums[left]]
                left += 1
            res += right - left + 1
        return res